README.txt
todoist-ng
    by James Britt / Neurogami
    http://www.neurogami.com

== DESCRIPTION:

Libs to make working with todoist.com Web API easier

== FEATURES/PROBLEMS:

* FIXME (list of features or problems)

== SYNOPSIS:

  FIXME (code sample of usage)

== REQUIREMENTS:

* FIXME (list of requirements)

== INSTALL:

* FIXME (sudo gem install, anything else)

== LICENSE:

(The MIT License)

Copyright (c) 2009 James Britt

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.0.0 / 2009-01-24

* 1 major enhancement
  * Birthday!


spec/spec_helper.rb

require File.expand_path(
    File.join(File.dirname(__FILE__), %w[.. lib todoist-ng]))

Spec::Runner.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
end

# EOF


spec/todoist-ng_spec.rb

require File.join(File.dirname(__FILE__), %w[spec_helper])

describe TodoistNg do
end

# EOF


lib/todoist-ng.rb
module Neurogami
  module Todoist
    class Setup
      # :stopdoc:
      LIBPATH = ::File.expand_path(::File.dirname(__FILE__)) + ::File::SEPARATOR
      PATH = ::File.dirname(LIBPATH) + ::File::SEPARATOR
      # :startdoc:

      # Returns the version string for the library.
      #

      # Returns the library path for the module. If any arguments are given,
      # they will be joined to the end of the libray path using
      # <tt>File.join</tt>.
      #
      def self.libpath( *args )
        args.empty? ? LIBPATH : ::File.join(LIBPATH, args.flatten)
      end

      # Returns the lpath for the module. If any arguments are given,
      # they will be joined to the end of the path using
      # <tt>File.join</tt>.
      #
      def self.path( *args )
        args.empty? ? PATH : ::File.join(PATH, args.flatten)
      end

      # Utility method used to rquire all files ending in .rb that lie in the
      # directory below this file that has the same name as the filename passed
      # in. Optionally, a specific _directory_ name can be passed in such that
      # the _filename_ does not have to be equivalent to the directory.
      #
      def self.require_all_libs_relative_to( fname, dir = nil )
        dir ||= ::File.basename(fname, '.*')
        search_me = ::File.expand_path(
                                       ::File.join(::File.dirname(fname), dir, '**', '*.rb'))

                                       Dir.glob(search_me).sort.each {|rb| require rb}
      end

    end  # module TodoistNg

  end
end

require 'version'

Neurogami::Todoist::Setup.require_all_libs_relative_to(__FILE__)



lib/todoist-ng/item.rb
require 'pp'

module Neurogami
  module Todoist

    # May want an Item class at some point ...
    class Item

      include Constants
=begin
 {"checked"=>0,
  "project_id"=>511874,
  "collapsed"=>0,
  "date_string"=>"mon",
  "priority"=>3,
  "id"=>3991694,
  "indent"=>1,
  "is_dst"=>0,
  "item_order"=>4,
  "chains"=>nil,
  "mm_offset"=>-420,
  "content"=>"test1232867261",
  "has_notifications"=>0,
  "due_date"=>"Mon Jan 26 23:59:59 2009",
  "in_history"=>0,
  "user_id"=>95496,
  "children"=>nil,
  "labels"=>[58599, 58593]},
=end
      attr_accessor :checked, :project_id, :collapsed, :date_string, :priority, :id, :indent, :is_dst
      attr_accessor :item_order, :chains, :mm_offset, :content, :has_notifications 
      attr_accessor :due_date, :in_history, :user_id, :children, :labels

      def initialize details
        details.each do |k,v|  
          # 'children' might be special as well ...
          next if k == 'labels'
          eval "@#{k} = '#{v}'"
        end

        @labels = details['labels'] # How do we want to work with these?
      end

    end

  end
end



__END__
https://todoist.com/API/help

Introduction
Todoist API can be used to hook Todoist into other applications. The API is work in progress and functionality will be added as people request it. So if you want some functionality, don't hesitate to contact me.
  Authentication
Every Todoist user has a special web service token, which can be found on the Preferences page. This token is used for authentication. The token represents the password, so don't distribute it, that would be like giving the password away.
  Solving cross domain security policy
You can't use AJAX to directly communicate with Todoist, this is due to browser security policy. You can solve this by communicating with Todoist using a script tag:
  var script = document.createElement('script');
script.type = 'text/javascript';
script.src = 'http://todoist.com/API/...&callback=callbackFunction';
document.getElementsByTagName('head')[0].appendChild(script);
The response data will look like this:
  callbackFunction({ JSON data here });
I.e. callbackFunction will be called.
  Projects and labels [top]
Get all projects
A request is sent to getProjects:
  http://todoist.com/API/getProjects?token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"user_id": 1, "name": "Test project", "color": 1, "collapsed": 0, "item_order": 1, "indent": 1, "cache_count": 4, "id": 22073},
    {"user_id": 1, "name": "Another test project", "color": 2, "collapsed": 0, "item_order": 2, "indent": 1, "cache_count": 0, "id": 22074},
    ...]
Get project
A request is sent to getProject with project_id:
  http://todoist.com/API/getProject?project_id=22073&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  {"user_id": 1L, "name": "Test project", "color": 1L, "collapsed": 0L, "item_order": 1L, "indent": 1L, "cache_count": 4L, "id": 22073L}
Get all labels
A request is sent to getLabels:
  http://todoist.com/API/getLabels?token=fb5f22601ec566e48083213f7573e908a7a272e5
Items [top]
Get uncompleted items
A request is sent to getCompletedItems with project_id:
  http://todoist.com/API/getUncompletedItems?project_id=22073&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"due_date": new Date("Sun Apr 29 23:59:59 2007"), "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 1, "faded": 0, "content": "By these things", "indent": 1, "project_id": 22073, "id": 210870, "checked": 0, "date_string": "29. Apr 2007"},
    {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 2, "faded": 0, "content": "Milk", "indent": 2, "project_id": 22073, "id": 210867, "checked": 0, "date_string": ""},
    ...]
Get completed items
A request is sent to getCompletedItems with project_id and optionally an offset:
  http://todoist.com/API/getCompletedItems?project_id=22073&offset=0&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 1, "priority": 1, "item_order": 2, "faded": 0, "content": "Fluffy ferret", "indent": 1, "project_id": 22073, "id": 210872, "checked": 1, "date_string": ""},
    {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 1, "priority": 1, "item_order": 1, "faded": 0, "content": "Test", "indent": 1, "project_id": 22073, "id": 210871, "checked": 1, "date_string": ""}
...]
Add an item to a project
A request is sent to addItem with project_id, content and optionally date_string, priority (int value from 1 to 4):
  http://todoist.com/API/addItem?content=Test&project_id=22073&priority=1&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 5, "faded": 0, "content": "Test", "indent": 1, "project_id": 22073, "id": 210873, "checked": 0, "date_string": null}
Update item
A request is sent to updateItem with id, content and optionally a date_string, priority (int value from 1 to 4):
  http://todoist.com/API/updateItem?id=210873&content=TestHello&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data of the item is returned:
  {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 5, "faded": 0, "content": "TestHello", "indent": 1, "project_id": 22073, "id": 210873, "checked": 0, "date_string": null}
Delete an item
A request is sent to deleteItems with items ids and project_id.
  Move items to the history
A request is sent to completeItems with item ids.
  http://todoist.com/API/completeItems?ids=[210873, 210873]&token=fb5f22601ec566e48083213f7573e908a7a272e5
Get items by id
A request is sent to getItemsById with item ids.
  http://todoist.com/API/getItemsById?ids=[210873, 210873]&token=fb5f22601ec566e48083213f7573e908a7a272e5
Query [top]
Queries
A request is sent to query with:

  * queries: A JSON string list of queries. Date format is 2007-4-29T10:59. Following queries are supported: viewall, overdue, p1, p2, p3

Example:
  http://todoist.com/API/query?queries=["2007-4-29T10:13","overdue","p1","p2"]&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"type": "date", "query": "2007-4-29T10:13", "data": [[...]]},
    {"type": "overdue", "data": [...]}, ...],




lib/todoist-ng/query.rb
require 'open-uri'
require 'cgi'
require 'json'
require 'pp'


module Neurogami
  module Todoist

    # May want an Item class at some point ...
    class Query
      include Constants

      # It would be nice to have a nicer UI than the user having to know this or that option
      # Following queries are supported: viewall, overdue, p1, p2, p3
      def self.run query_set, token
        #    * queries: A JSON string list of queries. Date format is 2007-4-29T10:59. 
        #    Following queries are supported: viewall, overdue, p1, p2, p3
        # Example:
        # http://todoist.com/API/query?queries=["2007-4-29T10:13","overdue","p1","p2"]&token=x2601ec5xxxxxxxxxxxxx573e908xxxa272e5 
        query_set.map!do |q|
            case q.class.to_s
            when 'Time'
                timestamp_to_querydate q
            when 'Symbol'
              q.to_s
            else
              q
            end
        end

        query =  CGI.escape(query_set.to_json)
        url = "#{BASE_URL}/query?queries=#{query};token=#{token}"
        results = JSON.parse(open( url ).read)

        # We get an array of hashes: type => data, where type is the query.
        #      [{"type"=>"overdue",
        #        "data"=>
        #          [{"checked"=>0,
        #            "project_id"=>501517,
        #            "collapsed"=>0,


        results.map do |result|
          { "type" => result["type"], "data" => result["data"].map{ |i| Item.new(i) } }     
        end
      end

      def self.timestamp_to_querydate time
        "#{time.year}-#{time.mon}-#{time.day}T#{time.hour}:#{time.min}"
      end
    end
  end
end



lib/todoist-ng/project.rb
require 'open-uri'
require 'cgi'
require 'json'
require 'pp'

module Neurogami
  module Todoist

    class Project

      include Constants
      
      @@token = 'NO_TOKEN!'
      @@labels  = nil
      @@projects = nil
      attr_accessor	:name, :id, :item_order, :cache_count, :color, :user_id, :items

      def self.token= t
        @@token = t
      end

      def initialize details 
        details.each { |k,v|  eval "@#{k} = '#{v}'"}
        @items = nil
      end



      def uncompleted_items
        # http://todoist.com/API/getUncompletedItems?project_id=22073&token=fb5f22601ec566e48083213f7573e908a7a272e5 
        begin
          items = JSON.parse(open( "#{BASE_URL}/getUncompletedItems?project_id=#{id};token=#{@@token}").read)
          items.map {|i| Item.new(i)}
        rescue Exception => e
          STDERR.puts " \n ERROR at #{__FILE__}:#{__LINE__}: Error getting uncompleted_items.\n #{e.inspect} "
          raise e
        end
      end


      def items reload=false
        return @items if @items && !reload
        @items = completed_items
        @items.concat(uncompleted_items)
        @items 
      end

      def completed_items offset = 0

        #Get completed items
        #A request is sent to getCompletedItems with project_id and optionally an offset:
        #http://todoist.com/API/getCompletedItems?project_id=22073&offset=0&token=fb5f22601ec566e48083213f7573e908a7a272e5 
        begin
          items = JSON.parse(open( "#{BASE_URL}/getCompletedItems?project_id=#{id};offset=#{offset};token=#{@@token}").read)
          items.map {|i| Item.new(i)}
        rescue Exception => e
          STDERR.puts " \n ERROR at #{__FILE__}:#{__LINE__}: Error getting completed_items\n #{e.inspect} "
          raise e
        end

      end

      def update_item item_id, content
        #        Update item
        #A request is sent to updateItem with id, content and optionally a date_string, priority (int value from 1 to 4):
        #http://todoist.com/API/updateItem?id=210873&content=TestHello&token=fb5f22601ec566e48083213f7573e908a7a272e5 
        begin
          JSON.parse(open( "#{BASE_URL}/updateItem?id=#{item_id};'content=#{content};token=#{@@token}").read)
        rescue Exception => e
          STDERR.puts " \n ERROR at #{__FILE__}:#{__LINE__}: Error calling update_item\n #{e.inspect} "
          raise e
        end

      end


      def add_item content, attributes = { :date_string => 'tomorrow', :priority => 2 }

        #    Add an item to a project
        #A request is sent to addItem with project_id, content and optionally date_string, priority (int value from 1 to 4):
        #http://todoist.com/API/addItem?content=Test&project_id=22073&priority=1&token=XXXXX22601ec56XXXX8083213fXXXXXX08a7a272e5
        #JSON data is returned:
        #{"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 5, "faded": 0, "content": "Test", "indent": 1, "project_id": 22073, "id": 210873, "checked": 0, "date_string": null}
        content.strip!
        raise "Cannot add an empty item!" if content.empty?
        if attributes[:labels]
          attributes[:labels]  =  attributes[:labels].split(',').map{ |l| " @#{l.strip}"}.join( ' ' )
          content <<  attributes[:labels]
          attributes.delete(:labels)
        end
        content = CGI.escape(content)


        attr_set = []

        attributes.each {|k,v|
          attr_set << "#{k.to_s}=#{CGI.escape(v.to_s)}"
        }  

        attr_set = attr_set.join(';')
        attr_set.strip!
        attr_set  = ";#{attr_set}" unless attr_set.empty? 
        url = "#{BASE_URL}/addItem?project_id=#{self.id};token=#{@@token};content=#{content}#{attr_set}"
        JSON.parse open(url).read
      end



      def delete_item item_id
        #        Delete an item
        #A request is sent to deleteItems with items ids and project_id.

        begin
          JSON.parse(open( "#{BASE_URL}/deleteItem?id=#{item_id};token=#{@@token}").read)
        rescue Exception => e
          STDERR.puts " \n ERROR at #{__FILE__}:#{__LINE__}: Error calling delete_item\n #{e.inspect} "
          raise e
        end

      end

      def self.project_named name
        projects.each do |prj|
          return prj if prj.name == name
        end
        nil
      end

      def self.get_projects reload=false
        return @@projects if @@projects && !reload
        begin
          results = JSON.parse(open( "#{BASE_URL}/getProjects?token=#{@@token}").read)
        rescue Exception => e
          STDERR.puts " \n ERROR at #{__FILE__}:#{__LINE__}: Error getting projects using url '#{BASE_URL}/getProjects?token=#{@@token}'\n #{e.inspect} "
          raise e
        end
        @@projects = results.map {|pr| Project.new(pr)} 
      end

      def self.projects(reload=false); get_projects(reload); end

      def self.get_project project_id
        result = JSON.parse(open( "#{BASE_URL}/getProject?project_id=#{project_id};token=#{@@token}").read)
        Project.new result 
      end

      def self.project(project_id); get_project(project_id); end

      def self.labels reload = false
        # http://todoist.com/API/getLabels?token=fb5f22601ec566e48083213f7573e908a7a272e5
        return @@labels if @@labels && !reload
        @@labels = JSON.parse(open( "#{BASE_URL}/getLabels?token=#{@@token}").read)
      end


      def self.add_item_to_project project_name, content, attributes = { :date_string => 'tomorrow @ 8am', :priority => 2 }
        project = project_named(project_name)
        project.add_item content, attributes
      end

    end

  end
end

__END__
https://todoist.com/API/help

Introduction
Todoist API can be used to hook Todoist into other applications. The API is work in progress and functionality will be added as people request it. So if you want some functionality, don't hesitate to contact me.
  Authentication
Every Todoist user has a special web service token, which can be found on the Preferences page. This token is used for authentication. The token represents the password, so don't distribute it, that would be like giving the password away.
  Solving cross domain security policy
You can't use AJAX to directly communicate with Todoist, this is due to browser security policy. You can solve this by communicating with Todoist using a script tag:
  var script = document.createElement('script');
script.type = 'text/javascript';
script.src = 'http://todoist.com/API/...&callback=callbackFunction';
document.getElementsByTagName('head')[0].appendChild(script);
The response data will look like this:
  callbackFunction({ JSON data here });
I.e. callbackFunction will be called.
  Projects and labels [top]
Get all projects
A request is sent to getProjects:
  http://todoist.com/API/getProjects?token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"user_id": 1, "name": "Test project", "color": 1, "collapsed": 0, "item_order": 1, "indent": 1, "cache_count": 4, "id": 22073},
    {"user_id": 1, "name": "Another test project", "color": 2, "collapsed": 0, "item_order": 2, "indent": 1, "cache_count": 0, "id": 22074},
    ...]
Get project
A request is sent to getProject with project_id:
  http://todoist.com/API/getProject?project_id=22073&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  {"user_id": 1L, "name": "Test project", "color": 1L, "collapsed": 0L, "item_order": 1L, "indent": 1L, "cache_count": 4L, "id": 22073L}
Get all labels
A request is sent to getLabels:
  http://todoist.com/API/getLabels?token=fb5f22601ec566e48083213f7573e908a7a272e5
Items [top]
Get uncompleted items
A request is sent to getCompletedItems with project_id:
  http://todoist.com/API/getUncompletedItems?project_id=22073&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"due_date": new Date("Sun Apr 29 23:59:59 2007"), "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 1, "faded": 0, "content": "By these things", "indent": 1, "project_id": 22073, "id": 210870, "checked": 0, "date_string": "29. Apr 2007"},
    {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 2, "faded": 0, "content": "Milk", "indent": 2, "project_id": 22073, "id": 210867, "checked": 0, "date_string": ""},
    ...]
Get completed items
A request is sent to getCompletedItems with project_id and optionally an offset:
  http://todoist.com/API/getCompletedItems?project_id=22073&offset=0&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 1, "priority": 1, "item_order": 2, "faded": 0, "content": "Fluffy ferret", "indent": 1, "project_id": 22073, "id": 210872, "checked": 1, "date_string": ""},
    {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 1, "priority": 1, "item_order": 1, "faded": 0, "content": "Test", "indent": 1, "project_id": 22073, "id": 210871, "checked": 1, "date_string": ""}
...]
Add an item to a project
A request is sent to addItem with project_id, content and optionally date_string, priority (int value from 1 to 4):
  http://todoist.com/API/addItem?content=Test&project_id=22073&priority=1&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 5, "faded": 0, "content": "Test", "indent": 1, "project_id": 22073, "id": 210873, "checked": 0, "date_string": null}
Update item
A request is sent to updateItem with id, content and optionally a date_string, priority (int value from 1 to 4):
  http://todoist.com/API/updateItem?id=210873&content=TestHello&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data of the item is returned:
  {"due_date": null, "user_id": 1, "collapsed": 0, "in_history": 0, "priority": 1, "item_order": 5, "faded": 0, "content": "TestHello", "indent": 1, "project_id": 22073, "id": 210873, "checked": 0, "date_string": null}
Delete an item
A request is sent to deleteItems with items ids and project_id.
  Move items to the history
A request is sent to completeItems with item ids.
  http://todoist.com/API/completeItems?ids=[210873, 210873]&token=fb5f22601ec566e48083213f7573e908a7a272e5
Get items by id
A request is sent to getItemsById with item ids.
  http://todoist.com/API/getItemsById?ids=[210873, 210873]&token=fb5f22601ec566e48083213f7573e908a7a272e5
Query [top]
Queries
A request is sent to query with:

  * queries: A JSON string list of queries. Date format is 2007-4-29T10:59. Following queries are supported: viewall, overdue, p1, p2, p3

Example:
  http://todoist.com/API/query?queries=["2007-4-29T10:13","overdue","p1","p2"]&token=fb5f22601ec566e48083213f7573e908a7a272e5
JSON data is returned:
  [{"type": "date", "query": "2007-4-29T10:13", "data": [[...]]},
    {"type": "overdue", "data": [...]}, ...],

